### PR TITLE
fix: add unique ID per Pattern usage

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
### Overview

Uses Lodash's `uniqueId` once per component usage to make pattern IDs unique.

This fixes a potential accessibility issue where multiple patterns on a page have the same `id` attribute.

### PR Checklist

- ~[ ] Related to designs:~
- [x] Related to JIRA ticket: GROW-2394
- [x] I have run this code to verify it works
- [x] This PR includes unit tests for the code change

### CODE REVIEW

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.4.38-alpha.e55f57.0<br/>@codecademy/gamut-labs@19.10.4-alpha.e55f57.0<br/>@codecademy/gamut@34.10.3-alpha.e55f57.0<br/>@codecademy/styleguide@41.10.8-alpha.e55f57.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://60d09b268fd0a787afae3b1f--gamut-preview.netlify.app>https://60d09b268fd0a787afae3b1f--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/60d09b268fd0a787afae3b1f>Deploy Logs</a></p>

File: packages/gamut/src/Pattern/index.tsx [hunk 1]
```diff
@@ -17,18 +17,15 @@ export interface PatternProps extends ComponentProps<typeof Box> {
   name: PatternName;
 }
 
-const Svg = styled.svg`
-  width: 100%;
-  height: 100%;
-`;
-
 export const Pattern: React.FC<PatternProps> = ({ name, ...props }) => {
+  const [id] = useState(() => uniqueId(name));
```
Comment (jakemhiller) [inline packages/gamut/src/Pattern/index.tsx:21]: Won't this be wonky in SSR? Or will it use the same indexes

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.4.38-alpha.6d6ca8.0<br/>@codecademy/gamut-labs@19.10.4-alpha.6d6ca8.0<br/>@codecademy/gamut@34.10.3-alpha.6d6ca8.0<br/>@codecademy/styleguide@41.10.8-alpha.6d6ca8.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://60d09e9cda44c908a89c0465--gamut-preview.netlify.app>https://60d09e9cda44c908a89c0465--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/60d09e9cda44c908a89c0465>Deploy Logs</a></p>

File: packages/gamut/src/Pattern/index.tsx (hunk 1 above)
Comment (JoshuaKGoldberg) [reply packages/gamut/src/Pattern/index.tsx:21]: It will be wonky but it won't be out of sync _within the component_.

Comment (JoshuaKGoldberg) [reply packages/gamut/src/Pattern/index.tsx:21]: Like, it should be that either the component re-hydrates or it doesn't. Either way, I'd think the `id` state would be consistent. I'll set up an integration branch on the monolith side to double check.

@zippyzow do you have an example of a place where you're seeing the double ID accessibility issue?

File: packages/gamut/src/Pattern/index.tsx [hunk 2]
```diff
@@ -17,18 +17,15 @@ export interface PatternProps extends ComponentProps<typeof Box> {
...
+
   return (
     <Box {...props}>
-      <Svg aria-hidden>
-        {defs(name)}
+      <Box aria-hidden as="svg" height="100%" width="100%">
```
Comment (MattSchiller) [inline packages/gamut/src/Pattern/index.tsx:25]: nice

**Review by MattSchiller — APPROVED** [review]
Love this solution, and agreed, I wonder if the SSR state should conflict here. V. curious to see what the answer is.

Like, the DOM's content (element IDs) should be different on rehydration, which is a server/client mismatch? I'm not sure is an issue? 🤔

File: packages/gamut/src/Pattern/index.tsx (hunk 1 above)
Comment (jakemhiller) [reply packages/gamut/src/Pattern/index.tsx:21]: the problem usually is that lodash's `uniqueId` uses the global namespace to keep track of id indexes, so b/c the renderer never clears that memory unless it restarts you eventually end up with like `pattern-12354567` in SSR. If you did come up with a global id system, you would maybe want to make it context based to make it local to the rendering context

Comment (zippyzow) [reply packages/gamut/src/Pattern/index.tsx:21]: @JoshuaKGoldberg for [this PR ](https://github.com/codecademy-engineering/Codecademy/pull/25292)

The new homepage design uses 2 `Pattern` components in the "How It Works" component. 

Screen Shot 2021-06-17 at 4 22 18 PM

**Comment by JoshuaKGoldberg** [comment]
Ok I've talked with a few people and I'm souring on this "intermediate" approach. The mismatch in hydration IDs is going to be painful.

I tried setting up a required `uniqueId` param but, ugh, that's actually a lot of plumbing.

Instead of trying to wrestle with some intermediate not-as-good solution I'm just going to go for the "right" solution: having a per-page registry.

**Comment by JoshuaKGoldberg** [comment]
The per-page registry is... hairy. I'll try this with a global React state.

**Comment by JoshuaKGoldberg** [comment]
Closing in favor of https://github.com/Codecademy/client-modules/pull/1863 after all ✨

File: packages/gamut/src/Pattern/index.tsx (hunk 1 above)
Comment (JoshuaKGoldberg) [reply packages/gamut/src/Pattern/index.tsx:21]: https://github.com/Codecademy/client-modules/pull/1863 ✨

**Pull request closed without merging**